pkg/minikube/machine/cache_binaries_test.go

```diff
@@ -0,0 +1,215 @@
+*/
+
+package machine
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"testing"
+
+	"k8s.io/minikube/pkg/minikube/assets"
+	"k8s.io/minikube/pkg/minikube/bootstrapper"
+	"k8s.io/minikube/pkg/minikube/command"
+)
+
+type copyFailRunner struct {
+	command.Runner
+}
+
+func (copyFailRunner) Copy(a assets.CopyableFile) error {
+	return fmt.Errorf("test error during copy file")
+}
+
+func newFakeCommandRunnerCopyFail() command.Runner {
+	return copyFailRunner{command.NewFakeCommandRunner()}
+}
+
+func TestCopyBinary(t *testing.T) {
+	var tc = []struct {
+		lastUpdateCheckFilePath string
+		src, dst, desc          string
+		err                     bool
+		runner                  command.Runner
+	}{
+		{
+			desc:   "not existing src",
+			dst:    "/tmp/testCopyBinary1",
+			src:    "/tmp/testCopyBinary2",
+			err:    true,
+			runner: command.NewFakeCommandRunner(),
+		},
+		{
+			desc:   "src /etc/hosts",
+			dst:    "/tmp/testCopyBinary1",
+			src:    "/etc/hosts",
+			err:    false,
+			runner: command.NewFakeCommandRunner(),
+		},
+		{
+			desc:   "existing src, copy fail",
+			dst:    "/etc/passwd",
+			src:    "/etc/hosts",
+			err:    true,
+			runner: newFakeCommandRunnerCopyFail(),
+		},
+	}
+	for _, test := range tc {
+		t.Run(test.desc, func(t *testing.T) {
+			err := CopyBinary(test.runner, test.src, test.dst)
+			if err != nil && !test.err {
+				t.Fatalf("Error %v expected but not occurred", err)
+			}
+			if err == nil && test.err {
+				t.Fatal("Unexpected error")
+			}
+		})
+	}
+}
+
+func TestCacheBinariesForBootstrapper(t *testing.T) {
+	var tc = []struct {
+		version, clusterBootstrapper string
+		err                          bool
+	}{
+		{
+			version:             "v1.16.0",
+			clusterBootstrapper: bootstrapper.BootstrapperTypeKubeadm,
+		},
+		{
+			version:             "invalid version",
+			clusterBootstrapper: bootstrapper.BootstrapperTypeKubeadm,
+			err:                 true,
+		},
+	}
+	for _, test := range tc {
+		t.Run(test.version, func(t *testing.T) {
+			err := CacheBinariesForBootstrapper(test.version, test.clusterBootstrapper)
+			if err != nil && !test.err {
+				t.Fatalf("Got unexpected error %v", err)
+			}
+			if err == nil && test.err {
+				t.Fatalf("Expected error but got %v", err)
+			}
+		})
+	}
+}
+func TestCacheBinary(t *testing.T) {
+	oldMinikubeHome := os.Getenv("MINIKUBE_HOME")
+	defer os.Setenv("MINIKUBE_HOME", oldMinikubeHome)
+
+	minikubeHome, err := ioutil.TempDir("/tmp", "")
+	if err != nil {
+		t.Fatalf("error during creating tmp dir: %v", err)
+	}
+	defer os.RemoveAll(minikubeHome)
+	noWritePermDir, err := ioutil.TempDir("/tmp", "")
+	if err != nil {
+		t.Fatalf("error during creating tmp dir: %v", err)
+	}
+	defer os.RemoveAll(noWritePermDir)
+	err = os.Chmod(noWritePermDir, 0000)
+	if err != nil {
+		t.Fatalf("error (%v) during changing permissions of dir %v", err, noWritePermDir)
+	}
+	noPermsDir := "/etc"
+	if runtime.GOOS == "windows" {
+		noPermsDir = "C:\\Windows\\System32"
```

Comment:
I don't have access to any windows machine, so this should work on windows, but It wasn't tested.